src/lib/y2storage/guided_proposal.rb

```diff
@@ -39,117 +39,135 @@ module Y2Storage
+      # @param settings [ProposalSettings]
+      # @param devicegraph [Devicegraph]
+      # @param disk_analyzer [DiskAnalyzer]
       #
-      # @return [GuidedProposal]
+      # @return [InitialGuidedProposal]
       def initial(settings: nil, devicegraph: nil, disk_analyzer: nil)
-        settings ||= ProposalSettings.new_for_current_product
-
-        strategy = initial_strategy(settings)
-
-        strategy.new.initial_proposal(
+        proposal = InitialGuidedProposal.new(
           settings:      settings,
           devicegraph:   devicegraph,
           disk_analyzer: disk_analyzer
         )
-      end
-
-    private
 
-      # Stragegy to create an initial proposal
-      #
-      # The strategy depends on the settings format.
-      #
-      # @see ProposalSettings#format
-      #
-      # @param settings [ProposalSettings]
-      # @return [Proposal::InitialStrategies::Legacy, Proposal::InitialStrategies::Ng]
-      def initial_strategy(settings)
-        if settings.format == ProposalSettings::LEGACY_FORMAT
-          Proposal::InitialStrategies::Legacy
-        else
-          Proposal::InitialStrategies::Ng
-        end
+        proposal.propose
+        proposal
+      rescue Y2Storage::Error
+        log.error("Initial proposal failed")
+        proposal
       end
     end
 
     # Constructor
     #
     # @param settings [ProposalSettings] if nil, default settings will be used
-    # @param devicegraph [Devicegraph] starting point. If nil, the probed
-    #   devicegraph will be used
+    # @param devicegraph [Devicegraph] starting point. If nil, the probed devicegraph
+    #   will be used.
     # @param disk_analyzer [DiskAnalyzer] by default, the method will create a new one
     #   based on the initial {devicegraph} or will use the one in {StorageManager} if
-    #   starting from probed (i.e. {devicegraph} argument is also missing)
+    #   starting from probed (i.e. {devicegraph} argument is also missing).
     def initialize(settings: nil, devicegraph: nil, disk_analyzer: nil)
       super(devicegraph: devicegraph, disk_analyzer: disk_analyzer)
+
       @settings = settings || ProposalSettings.new_for_current_product
     end
 
   private
 
+    # @return [ProposalSettings]
+    attr_writer :settings
+
+    # @return [Proposal::SpaceMaker]
+    attr_writer :space_maker
+
     # Calculates the proposal
     #
     # @raise [Error, NoDiskSpaceError] if there is no enough space to perform the installation
+    #
+    # @return [true]
     def calculate_proposal
-      settings.freeze
-
-      exception = nil
-      saved_root_device = populated_settings.root_device
-
-      groups = group_candidate_devices
-      target_sizes.product(groups).each do |target, devices|
-        # reset root_device, else #candidate_roots will just use it
-        populated_settings.root_device = saved_root_device
-        populated_settings.candidate_devices = devices
+      try_with_each_target_size
```

Comment:
Why this level of indirection? I mean having a method that is one-liner simply calling another method that, AFAICS, is not called anywhere else. :thinking:

Reply:
That `try_with_each_target_size` is used in derived class `InitialGuidedProposal`. The idea is that `GuidedProposal` only has to try with different sizes to calculate a proposal, and `InitialGuidedProposal` has to do more things before trying with different sizes.

Reply:
Never mind, I found why in the subclass.